File: src/main/java/com/study/csdojo/Problem11_LowestCommonAncestor.py
def lca(root, a, b):
    if not root:
        return None
    pathA = pathToX(root, a)
    pathB = pathToX(root, b)

    elemToReturn = None
    if pathA and pathB:
        while len(pathA) and len(pathB):
            elemA = pathA.pop()
            elemB = pathB.pop()
            if elemA.value == elemB.value:
                elemToReturn = elemA
            else:
                break
    return elemToReturn

def pathToX(root, x):
    if not root:
        return None
    if root.value == x:
        return [root]

    leftPath = pathToX(root.left, x)
    if leftPath:
        leftPath.append(root)
        return leftPath
    rightPath = pathToX(root.right, x)
    if rightPath:
        rightPath.append(root)
        return rightPath
    return None


class Node:
    def __init__(self, value, left=None, right=None):
        self.value = value
        self.left = left
        self.right = right

    def __str__(self):
        return str(self.value)

    # Overriding the equality operator.
    # This will be used for testing your solution.
    def __eq__(self, other):
        if type(other) is type(self):
            return self.value == other.value
        return False


# A function for creating a tree.
# Input:
# - mapping: a node-to-node mapping that shows how the tree should be constructed
# - head_value: the value that will be used for the head ndoe
# Output:
# - The head node of the resulting tree
def create_tree(mapping, head_value):
    head = Node(head_value)
    nodes = {head_value: head}
    for key, value in mapping.items():
        nodes[value[0]] = Node(value[0])
        nodes[value[1]] = Node(value[1])
    for key, value in mapping.items():
        nodes[key].left = nodes[value[0]]
        nodes[key].right = nodes[value[1]]
    return head


if __name__ == '__main__':
    # The mapping we're going to use for constructing a tree.
    # {0: [1, 2]} means that 0's left child is 1, and its right
    # child is 2.
    mapping1 = {0: [1, 2], 1: [3, 4], 2: [5, 6]}
    head1 = create_tree(mapping1, 0)
    # This tree is:
    # head1 = 0
    #        / \
    #       1   2
    #      /\   /\
    #     3  4 5  6


    mapping2 = {5: [1, 4], 1: [3, 8], 4: [9, 2], 3: [6, 7]}
    head2 = create_tree(mapping2, 5)
    # This tree is:
    #  head2 = 5
    #        /   \
    #       1     4
    #      /\    / \
    #     3  8  9  2
    #    /\
    #   6  7


    print("%s should return 0" % (lca(head1, 1, 5)))
    print("%s should return 1" % (lca(head1, 3, 1)))
    print("%s should return 1" % (lca(head1, 1, 4)))
    print("%s should return 0" % (lca(head1, 0, 5)))
    print("%s should return 5" % (lca(head2, 4, 7)))
    print("%s should return 3" % (lca(head2, 3, 3)))
    print("%s should return 1" % (lca(head2, 8, 7)))
    print("%s should return None (0 does not exist in the tree)" % (lca(head2, 3, 0)))

File: src/main/java/com/study/csdojo/Problem11_LowestCommonAncestor.java
package com.study.csdojo;

import java.util.LinkedList;

public class Problem11_LowestCommonAncestor {

    public TreeNode lca(TreeNode head, int a, int b) {
        if (head == null) {
            return null;
        }
        LinkedList<TreeNode> trackA = new LinkedList<>();
        if (!search(head, trackA, a)) {
            return null;
        }
        LinkedList<TreeNode> trackB = new LinkedList<>();
        if (!search(head, trackB, b)) {
            return null;
        }
        int i = -1;
        int minSize = Math.min(trackA.size(), trackB.size());
        TreeNode commonAncestor = null;
        while (minSize > 0) {
            TreeNode elemA = trackA.pop();
            TreeNode elemB = trackB.pop();
            if (elemA.value == elemB.value) {
                commonAncestor = elemA;
            } else {
                break;
            }
            minSize--;
        }
        return commonAncestor;
    }

    private boolean search(TreeNode head, LinkedList<TreeNode> track, int target) {
        if (head == null) {
            return false;
        }
        if (head.value == target || search(head.left, track, target) || search(head.right, track, target)) {
            track.push(head);
            return true;
        }
        return false;
    }

    static class TreeNode {
        private int value;
        public TreeNode left;
        public TreeNode right;

        public TreeNode(int value, TreeNode left, TreeNode right) {
            this.value = value;
            this.left = left;
            this.right = right;
        }
    }

}
